### PR TITLE
Extend suppression of config warnings 

### PR DESCRIPTION
**Pull request by alisihab**

### CODE REVIEW

File: core/src/main/resources/AppConstants.properties [hunk 1]
```diff
@@ -368,8 +368,16 @@ etag.ehcache.dir=${cache.dir}
 etag.ehcache.diskPersistent=false
 etag.ehcache.diskExpiryThreadIntervalSeconds=600
 
+# Below listed 4 suppression keys should be used with an adapter name to avoid losing any useful configuration warnings.
+# Example: warnings.suppress.transaction.{NameOfTheAdapter}
```
Comment (nielsm5) [inline core/src/main/resources/AppConstants.properties:372]: What do you mean with 'to avoid _losing_ any useful configuration warnings'

Also maybe add : warnings.suppress.transaction.{NameOfTheAdapter}**=true**, as just specifying the property isn't enough.

**Review by nielsm5 — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/configuration/ConfigurationWarnings.java [hunk 2]
```diff
@@ -36,7 +36,8 @@
 	public static final String SQL_INJECTION_SUPPRESS_KEY = "warnings.suppress.sqlInjections";
 	public static final String DEPRECATION_SUPPRESS_KEY="warnings.suppress.deprecated";
 	public static final String DEFAULT_VALUE_SUPPRESS_KEY = "warnings.suppress.defaultvalue";
-
+	public static final String TRANSACTION_SUPPRESS_KEY = "warnings.suppress.transaction";
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/configuration/ConfigurationWarnings.java:39]: Maybe it's time for this to become an Enum

File: core/src/main/resources/AppConstants.properties (hunk 1 above)
Comment (alisihab) [reply core/src/main/resources/AppConstants.properties:372]: transaction suppression key suppresses 3 different warnings all related to transactions. It is possible that a programmer sets the property to true then do some modification in the configuration and forget to set it to false and lose some of the config warnings.

File: core/src/main/java/nl/nn/adapterframework/extensions/mqtt/MqttListener.java [hunk 3]
```diff
@@ -171,4 +171,9 @@ public Message extractMessage(MqttMessage rawMessage, Map<String, Object> contex
 	@Override
 	public void afterMessageProcessed(PipeLineResult processResult, Object rawMessageOrWrapper, Map<String, Object> context) throws ListenerException {
 	}
+
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return super.getClassLoader();
+	}
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/extensions/mqtt/MqttListener.java:178]: Shouldn't you put this in the MqttFacade? It already has a configure, just no `IConfigurable` interface

File: core/src/main/java/nl/nn/adapterframework/http/WebServiceListener.java [hunk 4]
```diff
@@ -92,7 +93,7 @@ public void configure() throws ConfigurationException {
 
 		if (StringUtils.isEmpty(getServiceNamespaceURI()) && StringUtils.isEmpty(getAddress())) {
 			String msg = "calling webservices via de ServiceDispatcher_ServiceProxy is deprecated. Please specify an address or serviceNamespaceURI and modify the call accordingly";
-			ConfigurationWarnings.add(this, log, msg);
+			ConfigurationWarnings.add(this, log, msg, SuppressKeys.DEPRECATION_SUPPRESS_KEY, null);
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/WebServiceListener.java:96]: 👍

File: core/src/main/java/nl/nn/adapterframework/ftp/FtpSession.java [hunk 5]
```diff
@@ -68,6 +68,7 @@
  */
 public class FtpSession {
 	protected Logger log = LogUtil.getLogger(this);
+	private ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/ftp/FtpSession.java:71]: This class already has a configure method, If you add the `IConfigurable` interface here you don't have to add it on it's children.

**Review by nielsm5 — CHANGES_REQUESTED** [review]

File: core/src/main/resources/AppConstants.properties (hunk 1 above)
Comment (nielsm5) [reply core/src/main/resources/AppConstants.properties]: Ah I understand what you mean. I think it's best to add an attribute to the enums where you can specify if it works on global or on adapter level. In the case of `warnings.suppress.transaction`, `warnings.suppress.integrityCheck` and `warnings.suppress.sqlInjections` you never want to set this on a global level (as it's too dangerous like you mentioned). A developer must specifically suppress a warning for a given adapter, so future (new) adapters will get these warnings again.

File: core/src/main/java/nl/nn/adapterframework/ftp/FtpSession.java [hunk 6]
```diff
@@ -953,4 +957,19 @@ public boolean isKeyboardInteractive() {
 	public void setKeyboardInteractive(boolean keyboardInteractive) {
 		this.keyboardInteractive = keyboardInteractive;
 	}
+
+	@Override
+	@IbisDoc({"name of the listener", ""})
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/ftp/FtpSession.java:962]: In this case this has to be 'Name of the listener or sender'

File: core/src/main/java/nl/nn/adapterframework/ftp/FtpSession.java [hunk 7]
```diff
@@ -953,4 +957,19 @@ public boolean isKeyboardInteractive() {
...
+	}
+	@Override
+	public String getName() {
+		return name;
+	}
+
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return classLoader;
+	}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/ftp/FtpSession.java:974]: Please use ``@Getter`` annotation

File: core/src/main/java/nl/nn/adapterframework/http/IbisWebServiceSender.java [hunk 8]
```diff
@@ -40,6 +40,7 @@
  * @since 4.2
  */
 public class IbisWebServiceSender implements ISender, HasPhysicalDestination {
+	private ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/http/IbisWebServiceSender.java:43]: Please let this class extend SenderBase, and inherit name and classloader from that class

File: core/src/main/java/nl/nn/adapterframework/http/PushingListenerAdapter.java [hunk 9]
```diff
@@ -141,4 +142,9 @@ public void setRunning(boolean running) {
 		this.running = running;
 	}
 
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return classLoader;
+	}
+
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/http/PushingListenerAdapter.java:149]: Please use ``@Getter`` annotation

File: core/src/main/java/nl/nn/adapterframework/receivers/FileRecordListener.java [hunk 10]
```diff
@@ -284,6 +284,11 @@ public Map openThread() throws ListenerException {
 		return array.iterator();
 	}
 
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return classLoader;
+	}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/receivers/FileRecordListener.java:290]: Please use ``@Getter`` annotation

File: core/src/main/java/nl/nn/adapterframework/receivers/FileRecordListener.java [hunk 11]
```diff
@@ -366,4 +371,5 @@ public void setStoreFileNameInSessionKey(String storeFileNameInSessionKey) {
 	public String getStoreFileNameInSessionKey() {
 		return storeFileNameInSessionKey;
 	}
+
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/receivers/FileRecordListener.java:374]: unnecessary change

File: core/src/main/java/nl/nn/adapterframework/receivers/JavaListener.java [hunk 12]
```diff
@@ -216,6 +217,11 @@ public Message extractMessage(String rawMessage, Map<String,Object> context) thr
 		return new Message(rawMessage);
 	}
 
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return classLoader;
+	}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/receivers/JavaListener.java:223]: Please use ``@Getter`` annotation

File: core/src/main/java/nl/nn/adapterframework/receivers/ReceiverBase.java [hunk 13]
```diff
@@ -2305,4 +2306,8 @@ public String getHideMethod() {
 		return hideMethod;
 	}
 
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return classLoader;
+	}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/receivers/ReceiverBase.java:2312]: Please use ``@Getter`` annotation

File: core/src/main/java/nl/nn/adapterframework/senders/FileSender.java [hunk 14]
```diff
@@ -92,4 +92,9 @@ public ParameterList getParameterList() {
 		return paramList;
 	}
 
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return super.getClassLoader();
+	}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/senders/FileSender.java:98]: This shouldn't be necessary. Solve this in the super class

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/extensions/mqtt/MqttSender.java [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*
-   Copyright 2017 Integration Partners
+   Copyright 2017 - 2020 WeAreFrank!
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/extensions/mqtt/MqttSender.java:2]: Please remove this file from the PR, nothing has changed

File: core/src/main/java/nl/nn/adapterframework/ldap/LdapSender.java [hunk 16]
```diff
@@ -1347,4 +1347,9 @@ public void setReplyNotFound(boolean b) {
 	public boolean isReplyNotFound() {
 		return replyNotFound;
 	}
+
+	@Override
+	public ClassLoader getConfigurationClassLoader() {
+		return super.getClassLoader();
+	}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/ldap/LdapSender.java:1354]: This should not be necessaray

File: core/src/main/java/nl/nn/adapterframework/senders/FileSender.java [hunk 17]
```diff
@@ -1,5 +1,5 @@
 /*
-   Copyright 2013 Nationale-Nederlanden
+   Copyright 2013 Nationale-Nederlanden, 2020 WeAreFrank!
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/senders/FileSender.java:2]: Please remove this file from the PR, nothing has changed

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/ldap/LdapSender.java [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /*
-   Copyright 2013, 2020 Nationale-Nederlanden
+   Copyright 2013, 2020 Nationale-Nederlanden, 2020 WeAreFrank!
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/ldap/LdapSender.java:2]: Please remove this file from the PR, it has not changed

File: core/src/main/java/nl/nn/adapterframework/senders/FileSender.java [hunk 19]
```diff
@@ -91,5 +91,4 @@ public void addParameter(Parameter p) {
 	public ParameterList getParameterList() {
 		return paramList;
 	}
-
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/senders/FileSender.java:94]: Please remove this file from the PR, it has not changed

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

**Review by gvanbrakel — APPROVED** [review]

File: core/src/main/java/nl/nn/adapterframework/configuration/SuppressKeys.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+public enum SuppressKeys {
+
+	SQL_INJECTION_SUPPRESS_KEY("warnings.suppress.sqlInjections"),
+	DEPRECATION_SUPPRESS_KEY("warnings.suppress.deprecated"),
+	DEFAULT_VALUE_SUPPRESS_KEY("warnings.suppress.defaultvalue"),
+	TRANSACTION_SUPPRESS_KEY("warnings.suppress.transaction"),
+	INTEGRITY_CHECK_SUPPRESS_KEY("warnings.suppress.integrityCheck");
+
+	private @Getter String key;
+
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/configuration/SuppressKeys.java:35]: We suggest to add an option 'noGlobal', that, when set true, it inhibits suppression at the global level,

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

**Review by gvanbrakel — APPROVED** [review]

**Review by nielsm5 — CHANGES_REQUESTED** [review]

**Review by nielsm5 — APPROVED** [review]

File: core/src/main/java/nl/nn/adapterframework/configuration/ConfigurationWarnings.java [hunk 21]
```diff
@@ -33,9 +33,19 @@
...
+	public static void addGlobalWarning(Logger log, String msg, SuppressKeys suppressionKey, ClassLoader classLoader) {
+		if(suppressionKey == null) {
+			addGlobalWarning(log, msg, null);
+		} else if(!isSuppressed(suppressionKey, null, classLoader)) {
+			addGlobalWarning(log, msg, null);
+			// provide suppression hint as info
+			if(log.isInfoEnabled() && suppressionKey.isAllowGlobalSuppression()) {
+				String msgWithHint = msg + ". This warning can be suppressed by setting the property '"+suppressionKey.getKey()+"=true'";
+				log.info(msgWithHint);
+			}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/configuration/ConfigurationWarnings.java:46]: Now msg is logged twice: first without the hint in addGlobalWarning(), then here with the hint.
This should not happen. The message should be logged only once, (with or without hint, depending on log.isInfoEnabled())

File: core/src/main/java/nl/nn/adapterframework/configuration/ConfigurationWarnings.java [hunk 22]
```diff
@@ -61,9 +71,22 @@ public static void add(INamedObject object, Logger log, String message) {
...
+			if(log.isInfoEnabled()) {
+				if(adapter != null) {
+					String messageWithHint = "in adapter [" + adapter.getName() + "] " + message + ". This warning can be suppressed by setting the property '"+suppressionKey.getKey()+"."+adapter.getName()+"=true'";
+					if(suppressionKey.isAllowGlobalSuppression()) {
+						messageWithHint += ", or globally by setting the property '"+suppressionKey.getKey()+"=true'";
+					}
+					log.info(messageWithHint);
+				} else if(suppressionKey.isAllowGlobalSuppression()) {
+					String messageWithHint = message + "This warning can be suppressed globally by setting the property '"+suppressionKey.getKey()+"=true'";
+					log.info(messageWithHint);
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/configuration/ConfigurationWarnings.java:87]: Now msg is logged twice: first without the hint in add(), then here with the hint.
This should not happen. The message should be logged only once, (with or without hint, depending on log.isInfoEnabled())

File: core/src/main/java/nl/nn/adapterframework/http/IbisWebServiceSender.java [hunk 23]
```diff
@@ -39,9 +40,8 @@
  * @author Gerrit van Brakel
  * @since 4.2
  */
-public class IbisWebServiceSender implements ISender, HasPhysicalDestination {
+public class IbisWebServiceSender extends SenderBase implements ISender, HasPhysicalDestination {
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/http/IbisWebServiceSender.java:43]: When it extends SenderBase, it is not necessary to explicitly implement ISender, that comes with SenderBase

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

**Review by gvanbrakel — APPROVED** [review]

**Review by nielsm5 — APPROVED** [review]
